src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,769 @@
+                {
+                    expression = ((ParenthesizedExpressionSyntax)expression).Expression;
+                }
+
+                return expression.WithoutTrivia();
+            }
+
+            private static AnonymousObjectMemberDeclaratorSyntax CreateAnonymousObjectMemberDeclarator(string name)
+                => SyntaxFactory.AnonymousObjectMemberDeclarator(
+                    SyntaxFactory.NameEquals(SyntaxFactory.IdentifierName(name)), SyntaxFactory.IdentifierName(name));
+
+            private static StatementSyntax CreateJoinStatement(JoinClauseSyntax joinClause, StatementSyntax statement)
+            {
+                return SyntaxFactory.Block(
+                                SyntaxFactory.ForEachStatement(
+                                    joinClause.Type ?? VarNameIdentifier,
+                                    joinClause.Identifier,
+                                    joinClause.InExpression,
+                                    SyntaxFactory.Block(
+                                        SyntaxFactory.IfStatement(
+                                            SyntaxFactory.InvocationExpression(
+                                                SyntaxFactory.MemberAccessExpression(
+                                                    SyntaxKind.SimpleMemberAccessExpression,
+                                                    SyntaxFactory.PredefinedType(SyntaxFactory.Token(SyntaxKind.ObjectKeyword)),
+                                                    SyntaxFactory.IdentifierName(nameof(object.Equals))),
+                                                SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(
+                                                    new[] {
+                                                        SyntaxFactory.Argument(joinClause.LeftExpression),
+                                                        SyntaxFactory.Argument(joinClause.RightExpression.WithoutTrailingTrivia())}))),
+                                            statement))));
+            }
+
+            private class ProcessingState
+            {
+                public ImmutableArray<string> VariableNames { get; private set; }
+
+                // Each clause generates a method converting StatementSyntax into StatementSyntax.
+                public Stack<Func<StatementSyntax, StatementSyntax>> StatementProcessingMethods { get; }
+
+                public Func<ExpressionSyntax, StatementSyntax> LeafExpressionCreationMethod { get; }
+
+                public ProcessingState(Func<ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod)
+                {
+                    VariableNames = ImmutableArray<string>.Empty;
+                    StatementProcessingMethods = new Stack<Func<StatementSyntax, StatementSyntax>>();
+                    LeafExpressionCreationMethod = leafExpressionCreationMethod;
+                }
+
+                public void AddVariableName(string variableName) =>
```

Comment:
nit: ide style is => on next line indented for multiline line consturcts that use =>